index.bs

```diff
@@ -6021,29 +6067,45 @@ side=] of [=transform streams=].
  id="transform-stream-default-sink-abort-algorithm">TransformStreamDefaultSinkAbortAlgorithm(|stream|,
  |reason|)</dfn> performs the following steps:
 
- 1. Perform ! [$TransformStreamError$](|stream|, |reason|).
- 1. Return [=a promise resolved with=] undefined.
+ 1. Let |controller| be |stream|.[=TransformStream/[[controller]]=].
+ 1. If |controller|.[=TransformStreamDefaultController/[[finishPromise]]=] is not undefined, return
+    |controller|.[=TransformStreamDefaultController/[[finishPromise]]=].
+ 1. Let |readable| be |stream|.[=TransformStream/[[readable]]=].
+ 1. Let |controller|.[=TransformStreamDefaultController/[[finishPromise]]=] be a new promise.
+ 1. Let |cancelPromise| be the result of performing
+    |controller|.[=TransformStreamDefaultController/[[cancelAlgorithm]]=], passing |reason|.
+ 1. Perform ! [$TransformStreamDefaultControllerClearAlgorithms$](|controller|).
+ 1. [=React=] to |cancelPromise|:
+  1. If |cancelPromise| was fulfilled, then:
+   1. Perform ! [$ReadableStreamDefaultControllerError$](|readable|.[=ReadableStream/[[controller]]=], |reason|).
+   1. [=Resolve=] |controller|.[=TransformStreamDefaultController/[[finishPromise]]=] with undefined.
+  1. If |cancelPromise| was rejected with reason |r|, then:
+   1. Perform ! [$ReadableStreamDefaultControllerError$](|readable|.[=ReadableStream/[[controller]]=], |r|).
+   1. [=Reject=] |controller|.[=TransformStreamDefaultController/[[finishPromise]]=] with |r|.
+ 1. Return |controller|.[=TransformStreamDefaultController/[[finishPromise]]=].
 </div>
 
 <div algorithm>
  <dfn abstract-op lt="TransformStreamDefaultSinkCloseAlgorithm"
  id="transform-stream-default-sink-close-algorithm">TransformStreamDefaultSinkCloseAlgorithm(|stream|)</dfn>
  performs the following steps:
 
- 1. Let |readable| be |stream|.[=TransformStream/[[readable]]=].
  1. Let |controller| be |stream|.[=TransformStream/[[controller]]=].
+ 1. If |controller|.[=TransformStreamDefaultController/[[finishPromise]]=] is not undefined, return
+    |controller|.[=TransformStreamDefaultController/[[finishPromise]]=].
+ 1. Let |readable| be |stream|.[=TransformStream/[[readable]]=].
+ 1. Let |controller|.[=TransformStreamDefaultController/[[finishPromise]]=] be a new promise.
  1. Let |flushPromise| be the result of performing
     |controller|.[=TransformStreamDefaultController/[[flushAlgorithm]]=].
  1. Perform ! [$TransformStreamDefaultControllerClearAlgorithms$](|controller|).
- 1. Return the result of [=reacting=] to |flushPromise|:
+ 1. [=React=] to |flushPromise|:
   1. If |flushPromise| was fulfilled, then:
-   1. If |readable|.[=ReadableStream/[[state]]=] is "`errored`", throw
-      |readable|.[=ReadableStream/[[storedError]]=].
```

Comment:
Why was this check removed?

Reply:
Because _I think_, it is not possible for the readable to get into the errored state while this processing is happening.

Let me explain:

1. User calls `writable.close()`, returning promise 1
2. `WS#close` hook is called, invoking `TransformStreamDefaultSinkCloseAlgorithm`. The close algorithm returns a pending promise
3. User calls `readable.cancel()`, returning promise 2
4. `RS#cancel` hook is called, returning `finishPromise` (which is pending)
5. close algorithm pending promise is resolved
6. `finishPromise` gets resolved because of this (in `TransformStreamDefaultSinkCloseAlgorithm`)
7. promise 1 resolves
8. promise 2 resolves

The error passed to `readable.cancel()` is swallowed.

Because the `controller` of the RS is not exposed to the user, it is not possible to close the RS without going thru `cancelAlgorithm` in the TS.

This is inline with my comments above that all of `rs.cancel`, `ws.close` and `ws.abort` will all either resolve or reject (never one resolves one rejects), and that the first one always wins.

Reply:
> Because the `controller` of the RS is not exposed to the user, it is not possible to close the RS without going thru `cancelAlgorithm` in the TS.

You can error the readable side by calling `TransformStreamDefaultController.error()`. So I believe this check is trying to handle some sort of race condition between `writer.close()` and `controller.error()`.

Perhaps it's to handle the case where `flush()` called `controller.error()` instead of rejecting?
```javascript
new TransformStream({
  flush(controller) {
    controller.error(new Error("Boom!"));
  }
})
```

Reply:
I see! That's a fun one. I hadn't considered this case. I'll update and add a test

Reply:
Fixed in https://github.com/whatwg/streams/pull/1283/commits/40eadec8a733451e3819a16c31934fcb2a7987e9. Updated tests in https://github.com/web-platform-tests/wpt/pull/40453/commits/1f2f6c8912f414763e9ec177c8dfe316809a48c5